### PR TITLE
feat: implement DeleteSnapshot

### LINKED ISSUE #18: [feature] Support snapshot & restore
https://kubernetes-csi.github.io/docs/snapshot-restore-feature.html#snapshot--restore-feature

### PR DESCRIPTION
**Pull request by ZeroMagic**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
- implement DeleteSnapshot
- use `shareURL` to delete fileshare instead of `fileServiceClient`. Then we can use `x-ms-delete-snapshots=include` to delete all share snapshots with the share.
- enable `csi.ControllerServiceCapability_RPC_CREATE_DELETE_SNAPSHOT`
- change the way to run sanity test in CI

**Which issue(s) this PR fixes**:

Fixes #18 

**Special notes for your reviewer**:

**Release note**:
```
implement DeleteSnapshot
```

### CODE REVIEW

**Comment by ZeroMagic** [comment]
All sanity tests are passed including `CreateSnapshot` and `DeleteSnapshot`.

File: pkg/azurefile/azurefile.go [hunk 1]
```diff
@@ -268,3 +268,14 @@ func checkShareNameBeginAndEnd(fileShareName string) bool {
 
 	return false
 }
+
+// get snapshot name according to snapshot id, e.g.
+// input: "rg#f5713de20cde511e8ba4900#csivolumename#2019-08-22T07:17:53.0000000Z"
+// output: 2019-08-22T07:17:53.0000000Z
+func getSnapshot(id string) (string, error) {
+	segments := strings.Split(id, seperator)
+	if len(segments) < 4 {
```
Comment (andyzhangx) [inline pkg/azurefile/azurefile.go]: I think we don't allow one more `#` :
`if len(segments) != 4 {`

File: .travis.yml [hunk 2]
```diff
@@ -28,7 +28,9 @@ jobs:
       name: "sanity test"
       install:
         - sudo apt update && sudo apt install cifs-utils procps -y
-      script: make sanity-test
+      script: 
+        - make azurefile
```
[the remark this answers is not in the document]
Comment (ZeroMagic) [reply .travis.yml]: If we just use `sanity.Test(t, config)` to run sanity test, `skip` parameter can not be specify. So here we need `azurefileplugin` binary, then use `csi-sanity` binary to test.

File: pkg/azurefile/controllerserver.go [hunk 3]
```diff
@@ -111,6 +111,8 @@ func (d *Driver) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest)
 		return nil, err
 	}
 
+	klog.Errorf("create file share(%s) successfully", fileShareName)
```
[the remark this answers is not in the document]
Comment (ZeroMagic) [reply pkg/azurefile/controllerserver.go]: Oh, it is just for debug. I forget to delete it.

File: pkg/azurefile/azurefile.go (hunk 1 above)
Comment (ZeroMagic) [reply pkg/azurefile/azurefile.go]: Ok. The unit test has been added in `azurefile_test.go`

File: pkg/azurefile/controllerserver.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ZeroMagic) [reply pkg/azurefile/controllerserver.go]: done.

File: pkg/azurefile/controllerserver.go [hunk 4]
```diff
@@ -131,23 +133,19 @@ func (d *Driver) DeleteVolume(ctx context.Context, req *csi.DeleteVolumeRequest)
 	}
 
 	volumeID := req.VolumeId
-	resourceGroupName, accountName, fileShareName, err := getFileShareInfo(volumeID)
+	shareURL, err := d.getShareUrl(volumeID)
 	if err != nil {
-		klog.Errorf("getFileShareInfo(%s) in DeleteVolume failed with error: %v", volumeID, err)
+		// According to CSI Driver Sanity Tester, should succeed when an invalid volume id is used
```
[the remark this answers is not in the document]
Comment (ZeroMagic) [reply pkg/azurefile/controllerserver.go]: done.

File: pkg/azurefile/controllerserver.go [hunk 5]
```diff
@@ -264,7 +262,31 @@ func (d *Driver) CreateSnapshot(ctx context.Context, req *csi.CreateSnapshotRequ
 
 // DeleteSnapshot delete a snapshot (todo)
 func (d *Driver) DeleteSnapshot(ctx context.Context, req *csi.DeleteSnapshotRequest) (*csi.DeleteSnapshotResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "")
+	klog.V(2).Infof("DeleteSnapshot called with request %v", *req)
```
[the remark this answers is not in the document]
Comment (ZeroMagic) [reply pkg/azurefile/controllerserver.go]: done.

File: pkg/azurefile/controllerserver.go [hunk 6]
```diff
@@ -264,7 +262,31 @@ func (d *Driver) CreateSnapshot(ctx context.Context, req *csi.CreateSnapshotRequ
...
+
+	if len(req.SnapshotId) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Snapshot ID must be provided")
+	}
+
+	shareURL, err := d.getShareUrl(req.SnapshotId)
+	if err != nil {
+		// According to CSI Driver Sanity Tester, should succeed when an invalid snapshot id is used
+		klog.Errorf("failed to get share url with (%s): %v", req.SnapshotId, err)
```
[the remark this answers is not in the document]
Comment (ZeroMagic) [reply pkg/azurefile/controllerserver.go]: done.

File: test/sanity/run-test.sh [hunk 7]
```diff
@@ -16,30 +16,38 @@
...
-		${GO_BIN_PATH} test -v ./test/sanity/...
-	fi
+ endpoint="unix:///tmp/csi.sock"
+
+node="CSINode"
+if [ $# -gt 0 ]; then
+	node=$1
 fi
+
+ echo "being to run sanity test ..."
```
[the remark this answers is not in the document]
Comment (ZeroMagic) [reply test/sanity/run-test.sh]: done.

File: pkg/azurefile/azurefile.go (hunk 1 above)
Comment (ZeroMagic) [reply pkg/azurefile/azurefile.go]: done.

**Pull request merged**